drivers/apc_modbus.c

```diff
@@ -0,0 +1,1267 @@
+/*  apc_modbus.c - Driver for APC Modbus UPS
+ *  Copyright © 2023 Axel Gembe <axel@gembe.net>
+ *
+ *  This program is free software; you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation; either version 2 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program; if not, write to the Free Software
+ *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+ */
+
+#include "main.h"
+#if defined NUT_MODBUS_HAS_USB
+#include "nut_libusb.h"
+#include "libhid.h"
+#include "hidparser.h"
+#endif /* defined NUT_MODBUS_HAS_USB */
+#include "timehead.h"
+#include "nut_stdint.h"
+
+#include <ctype.h>
+#include <stdio.h>
+#include <time.h>
+
+#include <modbus.h>
+
+#define DRIVER_NAME "NUT APC Modbus driver"
+#define DRIVER_VERSION "0.01"
+
+#if defined NUT_MODBUS_HAS_USB
+
+#define APC_VENDORID 0x051d
+
+/* USB IDs device table */
+static usb_device_id_t apc_modbus_usb_device_table[] = {
+	{ USB_DEVICE(APC_VENDORID, 0x0003), NULL },
+
+	/* Terminating entry */
+	{ 0, 0, NULL }
+};
+
+#endif /* defined NUT_MODBUS_HAS_USB */
+
+/* Constants */
+static const uint8_t modbus_default_slave_id = 1;
+
+static const int modbus_rtu_default_baudrate = 9600;
+static const char modbus_rtu_default_parity = 'N';
+static const int modbus_rtu_default_databits = 8;
+static const int modbus_rtu_default_stopbits = 1;
+
+static const uint16_t modbus_tcp_default_port = 502;
+
+#if defined NUT_MODBUS_HAS_USB
+static const HIDNode_t modbus_rtu_usb_usage_rx = 0xff8600fc;
+static const HIDNode_t modbus_rtu_usb_usage_tx = 0xff8600fd;
+#endif /* defined NUT_MODBUS_HAS_USB */
+
+/* Variables */
+static modbus_t *modbus_ctx = NULL;
+#if defined NUT_MODBUS_HAS_USB
+static USBDevice_t usbdevice;
+static USBDeviceMatcher_t *reopen_matcher = NULL;
+static USBDeviceMatcher_t *regex_matcher = NULL;
+static int is_usb = 0;
+#endif /* defined NUT_MODBUS_HAS_USB */
+static int is_open = 0;
+static double power_nominal;
+static double realpower_nominal;
+
+/* Function declarations */
+static int _apc_modbus_read_inventory(void);
+
+/* driver description structure */
+upsdrv_info_t upsdrv_info = {
+	DRIVER_NAME,
+	DRIVER_VERSION,
+	"Axel Gembe <axel@gembe.net>\n",
+	DRV_BETA,
```

Comment:
Should we change `DRV_BETA` to `DRV_STABLE`? I've been using the driver for a while now and I'm quite confident. It is not complete though, but what ever is? I just checked and none of the drivers are `DRV_COMPLETE`

Maybe we should also make it version 1.0.

Reply:
Not sure... maybe after the commands get implemented? Like, being able to shut down/power cycle, or manage the beeper, is important :)

Reply:
Yea, I guess collecting some feedback before `DRIVER_STABLE` can't hurt either.